### ISSUE TITLE
Monotone_constraints in LightGbm

### ISSUE THREAD

**Issue by petterton** [dotnet/machinelearning#1651]
In my ML problem I get significantly better results from `LightGbm` by using the `monotone_constraints`  [parameter](https://github.com/Microsoft/LightGBM/blob/master/docs/Parameters.rst#monotone_constraints). I can not see that this is available through the ML.NET interface. Could this be added?

**Comment by najeeb-kazmi** [comment]
I agree this would be a useful addition to ML.NET. We provide a wrapper for LightGBM but this parameter is not exposed. I will file this for our triage team to review and prioritize.

**Comment by singlis** [comment]
Here are links from lightGBM for adding monotone constraints:
Issue filed here: https://github.com/Microsoft/LightGBM/issues/14
Committed here: https://github.com/Microsoft/LightGBM/pull/1314

The version of lightGBM we are using in ml.net is 2.2.1.1 -- need to confirm if this version contains the support for monotone_constraints.

**Comment by justinormont** [comment]
@daholste: Assuming I'm understanding this parameter correctly, this can also help with model stacking. Currently, LightGBM is allowed to map the output of the sub-models in the stack without the constraint that "as the sub-models score increases/decreases, so should the final score". Hence it could map as `f(x) = ( x < 3 ? 1.0 : (x < 4 ? 0.0 : 2.0 ))`. When the sub-models correlate well with the label, we likely would benefit from a monotonically increasing meta-model for stacking.

**Comment by glebuk** [comment]
Update the LightGBM and expose the arg.  Note if this fixes #1625  as well.

**Comment by justinormont** [comment]
Our current version is from 3mo ago; the monotone_constraints was added (at least) 8 mo ago. 

So our current version of LightGBM should work without updating.

**Comment by justinormont** [comment]
I think all or nothing is ok. 

I'd like to have control at the per column level, but I'm not sure how to make it user friendly. 

My specific use case is, I have a stacked model. The sub-models are rational, therefore I'd like them to be positive-monotone when combined to give the final score; I also want to feed some raw features to the final learner as this stacking method shows promise. To get the same results, I currently duplicate the raw features as `NegRawFeat = RawFeat * -1`, then `xf=Concat{Features:SubModelScores,NegRawFeat,RawFeat} tr=LogisticRegression{nn=+}`. This accomplishes the goal, though looses the slot names making feature importance difficult.

Two ways that are slightly user friendly:
* I suppose we could have three input columns: { `Features`, `PosFeatures`, `NegFeatures` }, where `Features` is the normal as current. Then users can concatenate the columns into `PosFeatures` which they want to be positive-monotone. I think inventing new input columns is confusion to users, so this is bad.
* I don't think it is possible, but another route would be to take in a single `Features` column, as current, then the user specifies the names of the column (within) the `Features` column that they want as positive/negative. I don't think it's possible to locate the slots within the `Features` column corresponding to the user's specified column.

**Comment by singlis** [comment]
After talking with Tom, this needs more thought through on how this can be made more user friendly. The PR that I currently have here: #2330 has the user specifying the constraints based upon indices. This was primarily to handle the way LightGBM works, but ML.Net does not work this way and using indices after concatenating columns into a Features column does not give a clear way to know which indices map to which features. 

Tom recommended to do something similar to how Categorical Features works  and somehow we manage the mapping of feature name to indices (such as in the metadata). This would allow the user to specify the constraints based upon a name (which would map to indices) rather than the specific indices. 

Also from talking with Tom, it sounds like this work can be done post v1.0 as it would not affect any API changes. 

My vote is to pause on this for now and reinvestigate post 1.0. 

@shauheen and @TomFinley  feel free to comment if you have anything additional.

**Comment by justinormont** [comment]
**TLDR;** For the moment, I'd be quite happy to have purely positive / purely negative for all slots.

---

I agree w/ @TomFinley. I have [similar concerns](https://github.com/dotnet/machinelearning/pull/2330#discussion_r252441134) about the usability. 

I would also like the style similar to what I think @TomFinley  is purposing, the Categorical Features style, is this similar to the second style on my list; I thought it would be _too hard_; but if Tom thinks it's doable it seems like a great longer term solution.
https://github.com/dotnet/machinelearning/issues/1651#issuecomment-456688022
> * I don't think it is possible, but another route would be to take in a single `Features` column, as current, then the user specifies the names of the column (within) the `Features` column that they want as positive/negative. I don't think it's possible to locate the slots within the `Features` column corresponding to the user's specified column. Perhaps match on slot names?

For the moment, I'd be quite happy to have purely positive / purely negative for all slots. This addresses AutoML team's ability to use it for model stacking: 

> https://github.com/dotnet/machinelearning/issues/1651#issuecomment-449535672, 
> > @daholste: Assuming I'm understanding this parameter correctly, this can also help with model stacking. Currently, LightGBM is allowed to map the output of the sub-models in the stack without the constraint that "as the sub-models score increases/decreases, so should the final score". Hence it could map as `f(x) = ( x < 3 ? 1.0 : (x < 4 ? 0.0 : 2.0 ))`. When the sub-models correlate well with the label, we likely would benefit from a monotonically increasing meta-model for stacking. 
>
> https://github.com/dotnet/machinelearning/issues/1651#issuecomment-456688022
> > My specific use case is, I have a stacked model. The sub-models are rational, therefore I'd like them to be positive-monotone when combined to give the final score; I also want to feed some raw features to the final learner as this stacking method shows promise. To get the same results, I currently duplicate the raw features as `NegRawFeat = RawFeat * -1`, then `xf=Concat{Features:SubModelScores,NegRawFeat,RawFeat} tr=LogisticRegression{nn=+}`. This accomplishes the goal (for LR, though not for LightGBM), though looses the slot names making feature importance difficult.

/cc @shauheen

**Comment by KyGurasz** [comment]
What's the status on this? Reading the thread it seems to me like we can use monotonic constraints with LightGBM, but only if we constrain all features in the same way? Is there any documentation?

**Comment by michaelgsharp** [comment]
@luisquintanilla are you aware of any documentation on this? Closing this issue as its been resolved by the PR, but please post any documentation here on it anyways.

**Comment by luisquintanilla** [comment]
@michaelgsharp we don't have docs that I'm aware of. If you can't point me to the PR that solved this issue or any related tests we can add this to our docs backlog
